py/src/braintrust/framework.py

```diff
@@ -500,6 +500,14 @@ def Eval(
         except RuntimeError:  # 'RuntimeError: There is no current event loop...'
             loop = None
 
+        base_experiment_name = None
+
+        data_iterator = evaluator.data
+        if inspect.isclass(data_iterator):
+            data_iterator = data_iterator()
```

Comment:
is it fine to create an instance here? it appears that this code is here to handle the `data=BaseExperiment` case in particular, in which case it seems fine.

Reply:
Hmm I don't think so. I don't think we can add the property that this is ok to be side-effect free all of a sudden.

Reply:
OK, makes sense. In that case, we probably need to create the instance here and pass it along in an optional arg to `run_evaluator` so it only occurs once

Reply:
Yeah although annoying, that feels like the right solution. _or_ maybe you could memoize it. Like place it into a wrapper of some sort, so that you can call it multiple times?

Reply:
what about simply calling it _before_ creating the `Evaluator` and assigning the result to the `.data` field?

Reply:
Does that affect when it's run eg. in an eval file vs today?

Reply:
^ Yeah that would run the data function during the lazy load instead of during the eval, which is no good. I changed it to just check whether it's an instance of `BaseExperiment`, but in the TS SDK things were a bit trickier.